# RPA (Robotic Process Automation)
* operating system automation
    * pyautogui
        * control all operating system controls/ pixels based automation
    * rpa
* browser based automation
    * selenium
* web-scraping
    * beautiful soup
    * scrapy

## pyautogui

In [2]:
!pip install pyautogui --upgrade --force

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 61.2/61.2 kB 162.3 kB/s eta 0:00:001m140.3 kB/s eta 0:00:01
  Installing build dependencies ... done
  Getting requirements to build wheel ... done
  Preparing metadata (pyproject.toml) ... done
  Using cached PyMsgBox-1.0.9-py3-none-any.whl
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 168.2/168.2 kB 1.7 MB/s eta 0:00:00 MB/s eta 0:00:01
  Preparing metadata (setup.py) ... done
  Installing build dependencies ... done
  Getting requirements to build wheel ... done
  Preparing metadata (pyproject.toml) ... done
  Using cached PyGetWindow-0.0.9-py3-none-any.whl
  Using cached MouseInfo-0.1.3-py3-none-any.whl
  Using cached python3_xlib-0.15-py3-none-any.whl
  Using cached PyRect-0.2.0-py2.py3-none-any.whl
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 3.4/3.4 MB 226.5 kB/s eta 0:00:00m eta 0:00:010:00:01
  Using cached pyperclip-1.8.2-py3-none-any.whl
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 76.8/76.8 kB 156.4 kB/s eta 0:00:00 kB/s eta 0:00

# we have to see screen resolution

* size of screen
* postion of cursor
* move your curson at any position within screen resoluation



In [16]:
import pyautogui
pyautogui.size()

Size(width=1360, height=768)

* top_left = x=0, y=0
* top_right = x=1360, y=0
* bottom_left = x=0, y=768
* bottom_right = x=1360, y=768

In [21]:
pyautogui.position()

Point(x=222, y=83)

# Now create our custome program

In [28]:
pyautogui.moveTo(x=333, y=45, duration=0.5)# move cursor
pyautogui.click()
pyautogui.click(x=222,y=83, duration=0.5)# click
pyautogui.write("Pakistan")# typing
pyautogui.hotkey('enter')# for short keys

In [31]:
pyautogui.hotkey('ctrl',"t")
pyautogui.write("Pakistan")# typing
pyautogui.hotkey('enter')# for short keys

In [27]:
pyautogui.hotkey('enter',"t")

In [12]:
pyautogui.moveTo(x=0, y=0)

In [15]:
pyautogui.moveTo(x=0, y=0,  duration=2)

# task1
1. open new tab
2. write youtube 
3. write in search box "It expert by sir qasim"
4. click channel link
5. click any play list
6. like all playlist videos

* same resolution
* same operating system